### PR TITLE
Token Refresh Offset

### PR DESCRIPTION
**Pull request by jianghaolu**
## Summary
We are moving the token refresh offset from `AccessToken` to `TokenCredential`, to keep the originality and correctness of the `expiresAt()` getter on `AccessToken`. 
On top of refreshing the token at an offset before expiry, we are also allowing the current non-expired token to be used if a proactive refresh fails. The proactive refresh will be attempted again next time some thread needs to get a token. To prevent too many refresh attempts, there will also be a timeout in place.

### Initial token acquisition
The initial token acquisition will be done by the first thread entering the `getToken()` and all other threads will subscribe to a publisher that will emit the token once acquired.

For the first thread, once the token is acquired, it's cached and sent to the publisher. The entire operation on this thread is atomic. 

### When a valid token is in cache
When multiple threads are reading from the cache, at any given time one thread will check the token validity and if any proactive token refresh should happen. If not, it simply falls back to the cached token with minimal performance hit.

### Proactive token refresh
When a thread is checking the validity of a token, `shouldRefresh` predicate may return true before the token expires. The token is considered "about to expire" for our customer and this thread is responsible for doing a proactive token refresh. 

If the refresh succeeds, the cached token will be updated. If the refresh fails, no error will be raised (will be logged) and the currently cached token will continue to be used. The next refresh attempt will not be allowed until 30 seconds later - the next thread coming in within 30 seconds will not attempt a proactive token refresh.

Likewise, these are all happening on one thread at any given time. All the others are simply getting the cached token emitted by the publisher.

### After token expires
If proactive token refreshes keep failing, or a token hasn't been checked until the currently cached token has expired, the next thread calling `getToken()` will be responsible for acquiring a new token. If there is a 30 seconds refresh timeout currently in place, since there are no more available tokens to fallback to, the thread will wait for the timeout to be over to attempt another token refresh.

If the refresh fails at this point, we have run out of available tokens, thus raising an error to the downstream.

### CODE REVIEW

File: sdk/core/azure-core/src/main/java/com/azure/core/credential/SimpleTokenCache.java [hunk 1]
```diff
@@ -22,23 +24,36 @@
...
+            t -> OffsetDateTime.now().isAfter(t.getExpiresAt().minus(TokenCredential.DEFAULT_TOKEN_REFRESH_OFFSET)));
+    }
+
+    /**
+     * Creates an instance of RefreshableTokenCredential with default scheme "Bearer".
+     *
+     * @param tokenSupplier a method to get a new token
+     * @param tokenExpired a method to check if the cached token is expired
+     */
+    public SimpleTokenCache(Supplier<Mono<AccessToken>> tokenSupplier, Function<AccessToken, Boolean> tokenExpired) {
```
Comment (anuchandy) [inline sdk/core/azure-core/src/main/java/com/azure/core/credential/SimpleTokenCache.java:45]: `Predicate<AccessToken>` could be considered

File: sdk/core/azure-core/src/main/java/com/azure/core/credential/AccessToken.java [hunk 2]
```diff
@@ -19,7 +19,7 @@
      */
     public AccessToken(String token, OffsetDateTime expiresAt) {
         this.token = token;
-        this.expiresAt = expiresAt.minusMinutes(2); // 2 minutes before token expires
+        this.expiresAt = expiresAt;
```
Comment (g2vinay) [inline sdk/core/azure-core/src/main/java/com/azure/core/credential/AccessToken.java:22]: This will be a breaking change for developers not using the Identity library and using this class directly from the core for their custom credentials.

Comment (jianghaolu) [reply sdk/core/azure-core/src/main/java/com/azure/core/credential/AccessToken.java:22]: I synced with @srnagar and he confirmed this is allowed in a minor version bump. It's not an API change but a behavioral change/fix. Users using this in conjunction with BearerTokenAuthenticationPolicy will experience the same behavior as the token cache now handles the 2 minutes default.

**Comment by JimSuplizio** [comment]
@jianghaolu I think you may need to rebase with master. The changes in version_client.txt that are in this PR are already in master but the versions in the [pom files in your fork](https://github.com/jianghaolu/azure-sdk-for-java/blob/tokenrefreshoffset/sdk/cosmos/azure-cosmos/pom.xml#L16) aren't going to match which is going to cause the analyze step to fail.

**Comment by jianghaolu** [comment]
Thanks @JimSuplizio pulling in master.

File: eng/versioning/version_client.txt [hunk 3]
```diff
@@ -63,6 +63,8 @@ com.microsoft.azure:azure-spring-boot-test-core;2.2.4;2.2.5-beta.1
 # Format;
 # unreleased_<groupId>:<artifactId>;dependency-version
 # note: The unreleased dependencies will not be manipulated with the automatic PR creation code.
+unreleased_com.azure:azure-core;1.6.0-beta.1
+unreleased_com.azure:azure-core-http-netty;1.6.0-beta.1
```
Comment (JimSuplizio) [inline eng/versioning/version_client.txt:67]: unreleased_com.azure:azure-core-http-netty has been added but isn't used anywhere

Comment (jianghaolu) [reply eng/versioning/version_client.txt:67]: Thanks! This is fixed now being used in azure-identity's pom.

File: sdk/core/azure-core/src/main/java/com/azure/core/credential/TokenCredential.java [hunk 4]
```diff
@@ -5,15 +5,31 @@
...
      * Asynchronously get a token for a given resource/audience.
      * @param request the details of the token request
      * @return a Publisher that emits a single access token
      */
     Mono<AccessToken> getToken(TokenRequestContext request);
+
+    /**
+     * The duration before the actual token expiry to refresh the token. Default is 2 minutes.
+     *
+     * @return the current offset for token refresh
```
Comment (JonathanGiles) [inline sdk/core/azure-core/src/main/java/com/azure/core/credential/TokenCredential.java:30]: This is a little hard to parse. Maybe something like "Returns a Duration value representing the amount of time to subtract from the token expiry time, whereupon attempts will be made to refresh the token. By default this will occur two minutes prior to the expiry of the token."

Comment (jianghaolu) [reply sdk/core/azure-core/src/main/java/com/azure/core/credential/TokenCredential.java]: This is moved to `TokenRefreshOptions` and the javadocs are updated there.

File: sdk/core/azure-core/src/main/java/com/azure/core/credential/TokenRefreshOptions.java [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    }
+
+    /**
+     * Specifies a Duration value representing the amount of time to wait between token refreshes. This is to prevent
+     * sending too many requests to the authentication service.
+     *
+     * @param tokenRefreshTimeout the duration value representing the amount of time to wait between token refreshes
+     * @return TokenRefreshOptions
+     */
+    public TokenRefreshOptions setTokenRefreshTimeout(Duration tokenRefreshTimeout) {
```
Comment (g2vinay) [inline sdk/core/azure-core/src/main/java/com/azure/core/credential/TokenRefreshOptions.java:35]: should this be setTokenRefreshRetryTimeout ?

TokenRefreshTimeout can be confused with TokenRefreshOffset

Comment (jianghaolu) [reply sdk/core/azure-core/src/main/java/com/azure/core/credential/TokenRefreshOptions.java]: Updated.

**Comment by jianghaolu** [comment]
Moved to https://github.com/Azure/azure-sdk-for-java/pull/13029
